src/famfs_lib.c

```diff
@@ -703,8 +703,8 @@ famfs_mmap_superblock_and_log_raw(const char *devname,
 		*logp = (struct famfs_log *)((u64)sb_buf + FAMFS_SUPERBLOCK_SIZE);
 
 	/* Invalidate the processor cache for the superblock and log */
-	invalidate_processor_cache(*logp, (*logp)->famfs_log_len);
-	invalidate_processor_cache(sb, sb->ts_log_offset);
+	invalidate_processor_cache(*logp, FAMFS_LOG_LEN);
+	invalidate_processor_cache(sb, FAMFS_SUPERBLOCK_SIZE);
```

Comment:
OK, I do see an issue, but I'm not sure it's the one you're thinking of (let me know)

The basic architecture is that the superblock is intended to be forever fixed at 2MiB - which is plenty, and makes the log 2MiB aligned. So the superblock size/length should be accessed from the FAMFS_SUPERBLOCK_SIZE macro. (FAMFS_LOG_OFFSET is a synonym, but the former is more correct).

The log is not intended to be fixed-size though (although until your work, we had not tried different sizes). But the presence of the log size in both the superblock and the log header was in anticipation of supporting different log sizes (though not log resizing - there are better ways to get the same result). The rationale for having the size in both the superblock and the log header is this: The superblock is authoritative in describing the location and size of the log, but the log must also contain everything you need to use it - even if you don't always have access to the superblock.

So I claim the correct way to reference the log size is either ```sb->ts_log_len``` if you have a copy or a reference to the superblock, or ```logp->famfs_log_len``` if the superblock is not handy. They are required to be the same in superblocks and logs with valid headers.

The problem with my code above is that it references ```logp->famfs_log_len``` when flushing/invalidating the log; if the log had stale cache lines loaded, ```famfs_log_len``` could be invalid - making the flush nonsense. 

Same issue for the superblock Although ```sb->ts_log_offset == FAMFS_SUPERBLOCK_SIZE```, my code above references the size to be invalidated from the cache line I'm trying to invalidate (and it should use FAMFS_SUPERBLOCK_SIZE). But I don't want to hard code FAMFS_LOG_LEN here because I don't want the size of the log compiled into the famfs userland code - I want it specified in the superblock.

I think this would be the correct approach:
```
	invalidate_processor_cache(sb, FAMFS_SUPERBLOCK_SIZE); /* This kicks the superblock out of the cache */
	if (famfs_check_super(sb) == 0) { /* This loads the superblock back into the cache */
		/* Superblock is valid; invalidate cache for log: */
		invalidate_processor_cache(*logp, sb->ts_log_len);
	} else {
		/* there is no log because there is no valid superblock */
	}
```
And this should all work even if the memory is not cache-coherent, and the cache is stale, and the log is a non-standard size.

I'm working on a branch/PR that is intended to correctly handle the log as a variable-sized entity, and support specifying the log size via ```mkfs.famfs```. I will try to get the cache handling right this time. I should have that up within a day or two, and would appreciate your feedback - hopefully it won't break you're work...

Thoughts?

Reply:
> OK, I do see an issue, but I'm not sure it's the one you're thinking of (let me know)

I think it's the same.  :)

> The problem with my code above is that it references `logp->famfs_log_len` when flushing/invalidating the log; if the log had stale cache lines loaded, `famfs_log_len` could be invalid - making the flush nonsense.
> 
> Same issue for the superblock Although `sb->ts_log_offset == FAMFS_SUPERBLOCK_SIZE`, my code above references the size to be invalidated from the cache line I'm trying to invalidate (and it should use FAMFS_SUPERBLOCK_SIZE). But I don't want to hard code FAMFS_LOG_LEN here because I don't want the size of the log compiled into the famfs userland code - I want it specified in the superblock.
> 
> I think this would be the correct approach:
> 
> ```
> 	invalidate_processor_cache(sb, FAMFS_SUPERBLOCK_SIZE); /* This kicks the superblock out of the cache */
> 	if (famfs_check_super(sb) == 0) { /* This loads the superblock back into the cache */
> 		/* Superblock is valid; invalidate cache for log: */
> 		invalidate_processor_cache(*logp, sb->ts_log_len);
> 	} else {
> 		/* there is no log because there is no valid superblock */
> 	}
> ```

Yes, under the design where the super block size is fixed at 2M, this processing is better.
I still have a little doubt about reason 2 I raised in previous apply, will operations like the following cause problems?
```
mkfs.famfs --logsize 512M /dev/dax0.0  // Suppose we add a parameter 'logsize'
<some operations>
mkfs.famfs -k /dev/dax0.0 // I want change logsize to 256M
mkfs.famfs --logsize 256M /dev/dax0.0 // In this step, sb->ts_log_len is 512M, and --logsize is 256M, 
                                        the function 'invalidate_processor_cache' will cause mkfs.famfs crash, 
                                        because it tries to flush the virtual address segment range that is not
                                        mmap mapped.
```

> 
> And this should all work even if the memory is not cache-coherent, and the cache is stale, and the log is a non-standard size.
> 
> I'm working on a branch/PR that is intended to correctly handle the log as a variable-sized entity, and support specifying the log size via `mkfs.famfs`. I will try to get the cache handling right this time. I should have that up within a day or two, and would appreciate your feedback - hopefully it won't break you're work...

This is great, I think it will be very useful to add a parameter to set the log size in mkfs.famfs. 👍 

> 
> Thoughts?